examples/java8/src/main/java/org/apache/beam/examples/advanced/subprocess/kernel/SubProcessKernel.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples.advanced.subprocess.kernel;
+
+import java.io.IOException;
+import java.lang.ProcessBuilder.Redirect;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import org.apache.beam.examples.advanced.subprocess.configuration.SubProcessConfiguration;
+import org.apache.beam.examples.advanced.subprocess.utils.CallingSubProcessUtils;
+import org.apache.beam.examples.advanced.subprocess.utils.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the process kernel which deals with exec of the subprocess.
+ * It also deals with all I/O.
+ */
+public class SubProcessKernel {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SubProcessKernel.class);
+
+  private static final int MAX_SIZE_COMMAND_LINE_ARGS = 128 * 1024;
+
+  SubProcessConfiguration configuration;
+  ProcessBuilder processBuilder;
+
+  private SubProcessKernel() {};
+
+  /**
+   * Creates the SubProcess Kernel ready for execution.
+   * Will deal with all input and outputs to the SubProcess
+   * @param options
+   * @param binaryName
+   */
+  public SubProcessKernel(SubProcessConfiguration options, String binaryName) {
+    this.configuration = options;
+    this.processBuilder = new ProcessBuilder(binaryName);
+  }
+
+  public List<String> exec(SubProcessCommandLineArgs commands) throws Exception {
+    try (CallingSubProcessUtils.Permit permit = new CallingSubProcessUtils.Permit(processBuilder.command().get(0))) {
+
+      List<String> results = null;
+
+      try (SubProcessIOFiles outputFiles = new SubProcessIOFiles(configuration.getWorkerPath())) {
+
+        try {
+          Process process = execBinary(processBuilder, commands, outputFiles);
+          results = collectProcessResults(process, processBuilder, outputFiles);
+        } catch (Exception ex) {
+          LOG.error("Error running executable ", ex);
+          throw (ex);
```

Comment:
If the subprocess is still running after this exception, then you'll leak it. Is it easy to .attempt to terminate the process here if still running?

Reply:
Good point will add as TODO and deal with in another PR